assets/js/src/mailpoet_url_factory.js

```diff
@@ -0,0 +1,27 @@
+const addReferralId = (url, referralId) => {
+  if (!referralId) {
+    return url;
+  }
+  const refUrl = new URL(url);
+  refUrl.searchParams.set('ref', referralId);
+  return refUrl.toString();
+};
+
+const MailPoetUrlFactory = referralId => ({
+  getFreePlanUrl: (utmCampaign = null, utmMedium = null) => {
+    const url = new URL(addReferralId('https://www.mailpoet.com/free-plan/?utm_source=plugin', referralId));
```

Comment:
This doesn't look very nice - you create a URL string, parse it to URL object (in `addReferralId`), modify some parameters, output it to string and then parse it again.

This should work differently - i.e. adding a ("private") method `getUrl` that would return URL object with base URL and extending on that (by adding path and params). Above that, such method could already register the referral ID parameter.